server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java

```diff
@@ -137,11 +138,11 @@ public void close() {
                         chunkConsumer.streamClose();
                     }
                 });
+                action.accept(channel);
+            } else {
+                action.accept(channel);
+                request.getHttpRequest().release();
```

Comment:
IMO this is too far away from the point where we acquired this ref, and there's a bunch of other branches that don't release it like this. I think it's ok because `.release()` is a run-once thing, and we call it again when sending the response, but still this kind of trickiness is where bugs lurk.

We discussed the potential for doing this release directly in the Netty layers, maybe `Netty4HttpPipeliningHandler`, but I guess from the change as it stands that this didn't work because we need to retain it for longer when it's a streamed request. But could we always `.release()` the request in `Netty4HttpPipeliningHandler` and `.retain()` it in the streaming case to extend its lifecycle?

Reply:
I dug into this a bit more and it seems feasible but it's more invasive than I expected. Let's not do it in this PR, we can address it in a follow-up.

Reply:
I tried few quick changes to make acquire and release in netty stack, but it does not work well with our current implementation. The ES HttpRequest is a Closable rather than RefCounted. `close()` is idempotent and can happen only once. My attempts to change ref counts end up with either leak or early closure.

To align ES HttpRequest to it's RefCounted content we should change close behaviour and make request truly ref counted.